impeller/entity/geometry/stroke_path_geometry.cc

```diff
@@ -574,27 +557,43 @@ GeometryResult StrokePathGeometry::GetPositionBuffer(
     return {};
   }
 
-  Scalar min_size =
-      (pass.GetSampleCount() == SampleCount::kCount4 ? kMinStrokeSizeMSAA
-                                                     : kMinStrokeSize) /
-      max_basis;
+  Scalar min_size = kMinStrokeSize / max_basis;
   Scalar stroke_width = std::max(stroke_width_, min_size);
-
+  bool is_hairline = stroke_width_ <= min_size;
   auto& host_buffer = renderer.GetTransientsBuffer();
-  auto scale = entity.GetTransform().GetMaxBasisLength();
+
+  // This is a harline stroke and can be drawn directly with line primitives,
+  // which avoids extra tessellation work, cap/joins, and overdraw prevention.
+  if (is_hairline && path_.IsSingleContour()) {
+    // TODO(jonahwilliams): this could apply to multi contour paths if we add
+    // support for primitive restart.
+    auto vertex_buffer = renderer.GetTessellator()->TessellateConvex(
+        path_, host_buffer, max_basis, /*line_strip=*/true);
+    return GeometryResult{
+        .type = PrimitiveType::kLineStrip,
+        .vertex_buffer = vertex_buffer,
+        .transform = entity.GetShaderTransform(pass),
+    };
+  }
+  // For harline strokes that must be tessellated, switch to the cheaper
+  // cap and joins since they will be barely noticable at pixel scale.
+  Join join = stroke_join_;
+  Cap cap = stroke_cap_;
+  if (is_hairline) {
+    join = Join::kBevel;
+    cap = Cap::kButt;
```

Comment:
I fear we might miss pixels/MSAA-samples with these. Round is really the one you want to avoid as that will attempt to generate a tesselated circle. Square caps shouldn't be a problem as it is just one additional vector add. Miter is more than bevel, but by how much?

Reply:
The caps/joins are made of triangles, but this change makes hairline strokes visible via line strip primitives. We can't draw the cap/join with connected points, and its not clear that we need to as they'd certainly be subpixel already.

Reply:
butt cap are 1 line width shorter than square caps. Bevel might miss any samples on the vertex that the true outline of a miter or round join would capture. These could produce squares that are missing corners depending on the sub-pixel registration of the line.

Reply:
hmm, I could bump the ends out by about that much then? So that would be add 1px to each end for square/round and nothing for butt?

Reply:
I'm just going to remove this bit, its not really that important.